functions/src/game.ts

```diff
@@ -30,11 +31,140 @@ export function generateDeck() {
   return deck;
 }
 
+/** Check if three cards form a set. */
+export function checkSet(a: string, b: string, c: string) {
+  for (let i = 0; i < 4; i++) {
+    if ((a.charCodeAt(i) + b.charCodeAt(i) + c.charCodeAt(i)) % 3 !== 0)
+      return false;
+  }
+  return true;
+}
+
+/** Check if four cards form an ultraset */
+export function checkSetUltra(a: string, b: string, c: string, d: string) {
+  for (let i = 1; i < 4; i++) {
+    let ok = true;
+    for (let j = 0; j < 4; j++) {
+      const mod = [0, 0];
+      for (let k = 0; k < 4; k++) {
+        if (k === 0 || k === i) mod[0] += arguments[k].charCodeAt(j);
+        else mod[1] += arguments[k].charCodeAt(j);
+      }
+      if (mod[0] % 3 !== mod[1] % 3) ok = false;
+    }
+    if (ok) {
+      const ret = [a, arguments[i]];
+      for (let j = 1; j < 4; j++) {
+        if (j !== i) ret.push(arguments[j]);
+      }
+      return ret;
+    }
+  }
+  return null;
+}
+
+/** Find a set in an unordered collection of cards, if any, depending on mode. */
+export function findSet(
+  deck: string[],
+  gameMode = "normal",
+  old: string[] = []
```

Comment:
I'm not sure if I understand, since later I do use it in `for (let k of old)`, which might be bad if `old` is `undefined`. Are you saying that it's fine since I only use `old` when `gameMode === "setchain"` and we would always pass in an array in that case anyways?

Reply:
I don't believe the default value is used; you could set it to `["foo", "bar"]` and nothing would change, right?